sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java

```diff
@@ -454,50 +457,87 @@ private synchronized void triggerIngestion(final @NonNull GroupState groupState)
-        if (batchId == null) {
-            return;
-        }
-        AppCenterLog.debug(LOG_TAG, "ingestLogs(" + groupState.mName + "," + batchId + ") pendingLogCount=" + groupState.mPendingLogCount);
+            /* Get a batch from Persistence. */
+            final List<Log> batch = new ArrayList<>(maxFetch);
+            final int stateSnapshot = mCurrentState;
+            final String batchId = mPersistence.getLogs(groupState.mName, groupState.mPausedTargetKeys, maxFetch, batch, startTime, endTime);
 
-        /* Call group listener before sending logs to ingestion service. */
-        if (groupState.mListener != null) {
-            for (Log log : batch) {
-                groupState.mListener.onBeforeSending(log);
-            }
-        }
+            /* Decrement counter. */
+            groupState.mPendingLogCount -= maxFetch;
 
-        /* Remember this batch. */
-        groupState.mSendingBatches.put(batchId, batch);
+            /* If there are no logs to send. */
+            if (batchId == null) {
 
-        /*
-         * Due to bug on old Android versions (verified on 4.0.4),
-         * if we start an async task from here, i.e. the async handler thread,
-         * we end up with AsyncTask configured with the wrong Handler to use for onPostExecute
-         * instead of using main thread as advertised in Javadoc (and its a static field there).
-         *
-         * Our SDK guards against an application that would make a first async task in non UI
-         * thread before SDK is initialized, but we should also avoid corrupting AsyncTask
-         * with our wrong handler to avoid creating bugs in the application code since we are
-         * a library.
-         *
-         * So make sure we execute the async task from UI thread to avoid any issue.
-         */
-        HandlerUtils.runOnUiThread(new Runnable() {
+                /* Remove oldest token if there are no more logs. */
+                if (!iterator.hasPrevious() &&
+                        authTokenStorage != null && endTime != null &&
+                        mPersistence.countLogs(endTime) == 0) {
+                    authTokenStorage.removeToken(authToken);
+                }
+                continue;
+            }
+            AppCenterLog.debug(LOG_TAG, "ingestLogs(" + groupState.mName + "," + batchId + ") pendingLogCount=" + groupState.mPendingLogCount);
 
-            @Override
-            public void run() {
-                sendLogs(groupState, stateSnapshot, batch, batchId);
+            /* Call group listener before sending logs to ingestion service. */
+            if (groupState.mListener != null) {
+                for (Log log : batch) {
+                    groupState.mListener.onBeforeSending(log);
+                }
             }
-        });
+
+            /* Remember this batch. */
+            groupState.mSendingBatches.put(batchId, batch);
+
+            /*
+             * Due to bug on old Android versions (verified on 4.0.4),
```

Comment:
Unrelated FYI (no action required in this PR): not sure when they fixed it but we dropped support for API level 15.